src/ocean/meta/traits/Basic.d

```diff
@@ -0,0 +1,508 @@
+
+    static assert (!isPointerType!(Bacon));
+
+    // function pointer is a pointer, but delegate is not:
+    void foo () {}
+    static void bar () {}
+    static assert (!isPointerType!(typeof(&foo)));
+    static assert ( isPointerType!(typeof(&bar)));
+}
+
+/*******************************************************************************
+
+    Params:
+        T = type to check
+
+    Returns:
+        `true` if `T` is a struct, class, interface or union
+
+*******************************************************************************/
+
+template isAggregateType ( T )
+{
+    const isAggregateType =
+           is(T == class)
+        || is(T == interface)
+        || is(T == struct)
+        || is(T == union);
+}
+
+///
+unittest
+{
+    struct S { int x; }
+    union U { int x; double y; }
+
+    static assert ( isAggregateType!(S));
+    static assert ( isAggregateType!(U));
+    static assert (!isAggregateType!(S[2]));
+}
+
+/*******************************************************************************
+
+    Params:
+        T = type to check
+
+    Returns:
+        `true` if T is a function, function pointer or delegate
+
+*******************************************************************************/
+
+template isFunctionType ( T )
+{
+    const bool isFunctionType =
+           is(T == function)
+        || is(typeof(*T.init) == function)
+        || is(T == delegate);
+}
+
+///
+unittest
+{
+    void foo1() { }
+    auto foo2 = () { };
+    static void foo3() { }
+
+    static assert (isFunctionType!(typeof(&foo1)));
+    static assert (isFunctionType!(typeof(foo2)));
+    static assert (isFunctionType!(typeof(&foo3)));
+    static assert (isFunctionType!(typeof(foo3)));
+}
+
+unittest
+{
+    static real func(ref int) { return 0; }
+    void nestedFunc() { }
+    class C
+    {
+        real method(ref int) { return 0; }
+    }
+    auto c = new C;
+    auto fp = &func;
+    auto dg = &c.method;
+    real val;
+
+    static assert ( isFunctionType!(typeof(func)));
+    static assert ( isFunctionType!(typeof(nestedFunc)));
+    static assert ( isFunctionType!(typeof(C.method)));
+    static assert ( isFunctionType!(typeof(fp)));
+    static assert ( isFunctionType!(typeof(dg)));
+    static assert ( isFunctionType!(real function(ref int)));
+    static assert ( isFunctionType!(real delegate(ref int)));
+    static assert ( isFunctionType!(typeof((int a) { return a; })));
+
+    static assert (!isFunctionType!(int));
+    static assert (!isFunctionType!(typeof(val)));
+}
+
+/*******************************************************************************
+
+    Params:
+        T = type to check
+
+    Returns:
+        `true` if T is a function, function pointer or delegate or callable
+        aggregate
+
+*******************************************************************************/
+
+template isCallableType ( T )
+{
+    const bool isCallableType =
+           isFunctionType!(T)
+        || is(typeof(T.opCall));
```

Comment:
Minor point/out of interest: Wouldn't it be better to check if `typeof(&T.init.opCall)` is a function or delegate? (I could imagine that could be a hassle, though.)